k9mail/src/main/java/com/fsck/k9/service/FileProviderDeferredFileOutputStream.java

```diff
@@ -0,0 +1,63 @@
+package com.fsck.k9.service;
+
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import org.apache.commons.io.output.DeferredFileOutputStream;
+import org.apache.commons.io.output.ThresholdingOutputStream;
+
+
+/**
+ * This OutputStream is modelled after apache commons' {@link DeferredFileOutputStream},
+ * but uses a {@link FileProviderInterface} instead of directly generating temporary files
+ * itself.
+ */
+public class FileProviderDeferredFileOutputStream extends ThresholdingOutputStream {
+
+
+    private ByteArrayOutputStream memoryOutputStream;
+    private OutputStream currentOutputStream;
+    private File outputFile;
+    private final FileProviderInterface fileProviderInterface;
+
+
+    public FileProviderDeferredFileOutputStream(int threshold, FileProviderInterface fileProviderInterface) {
```

Comment:
Instead of `FileProviderInterface` pass in a `FileFactory` that only knows how to create a file. Again, not mixing this with the file provider stuff.